File: src/main/java/com/ultreon/mods/screenshotmanager/client/gui/screens/ScreenshotsScreen.java
package com.ultreon.mods.screenshotmanager.client.gui.screens;

import com.mojang.blaze3d.platform.InputConstants;
import com.mojang.blaze3d.platform.NativeImage;
import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.PoseStack;
import com.ultreon.mods.guilib.client.gui.screen.FullscreenRenderScreen;
import com.ultreon.mods.guilib.client.gui.widget.ToolbarButton;
import com.ultreon.mods.screenshotmanager.MainMod;
import com.ultreon.mods.screenshotmanager.client.Screenshot;
import com.ultreon.mods.screenshotmanager.client.ScreenshotCache;
import com.ultreon.mods.screenshotmanager.client.ScreenshotData;
import com.ultreon.mods.screenshotmanager.common.FloatSize;
import com.ultreon.mods.screenshotmanager.common.KeyboardHelper;
import com.ultreon.mods.screenshotmanager.common.Resizer;
import com.ultreon.mods.screenshotmanager.text.CommonTexts;
import lombok.SneakyThrows;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.client.renderer.texture.AbstractTexture;
import net.minecraft.client.renderer.texture.DynamicTexture;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.util.*;
import java.util.concurrent.atomic.AtomicBoolean;

public class ScreenshotsScreen extends FullscreenRenderScreen {
    // No getter / setter.
    private final List<File> files = new ArrayList<>();
    private final List<Screenshot> screenshots = new ArrayList<>();

    // Getter only.
    private final Screen backScreen;
    private Screenshot currentScreenshot;

    // Getter & setter.
    private int index = 0;
    private Thread loadThread;
    private int currentIndex = -1;
    private int total;
    private int loaded;

    private boolean loading = false;

    public ScreenshotsScreen(Component title, Screen backScreen) {
        super(title);
        this.backScreen = backScreen;

        addToolbarItem(new ToolbarButton(0, 0, 50, CommonTexts.PREV, toolbarButton -> prevShot()));
        addToolbarItem(new ToolbarButton(0, 0, 50, CommonTexts.NEXT, toolbarButton -> nextShot()));

        this.reload();
    }

    private void reload() {
        loading = true;
        File dir = new File(Minecraft.getInstance().gameDirectory, "screenshots");
        if (dir.exists()) {
            this.files.addAll(Arrays.asList(Objects.requireNonNull(dir.listFiles())));
        }
        this.total = this.files.size();
        this.index = 0;

        this.screenshots.clear();

        this.loadThread = new Thread(this::loadShots, "ScreenshotLoader");
        this.loadThread.start();
    }

    @SuppressWarnings({"BusyWait", "ConstantConditions"})
    @SneakyThrows
    private void loadShots() {
        this.loaded = 0;

        MainMod.LOGGER.info("Refreshing screenshot cache.");
        AtomicBoolean active = new AtomicBoolean(true);
        for (File file : this.files) {
            active.set(true);
            RenderSystem.recordRenderCall(() -> {
                AbstractTexture texture;
                ScreenshotData data;
                @Nullable ResourceLocation location;

                if (!ScreenshotCache.get().isCached(file)) {
                    location = new ResourceLocation(MainMod.MOD_ID, "screenshots_screen/" + file.getName().toLowerCase(Locale.ROOT).replaceAll("[^a-z0-9/._-]", "_"));
                    AbstractTexture tex = Minecraft.getInstance().getTextureManager().getTexture(location, null);

                    if (tex == null) {
                        DynamicTexture dynamicTexture = this.loadTexture(location, file);
                        texture = dynamicTexture;
                        data = ScreenshotCache.get().cache(file, dynamicTexture);

                        if (texture == null) {
                            location = null;
                        }
                    } else {
                        texture = tex;
                        data = ScreenshotCache.get().get(file);
                    }
                } else {
                    data = ScreenshotCache.get().get(file);
                    location = new ResourceLocation(MainMod.MOD_ID, "screenshots_screen/" + file.getName().toLowerCase(Locale.ROOT).replaceAll("[^a-z0-9/._-]", "_"));
                    texture = Minecraft.getInstance().getTextureManager().getTexture(location, null);

                }
                Screenshot screenshot = new Screenshot(file, texture, location, data);
                screenshots.add(screenshot);
                this.loaded++;
                active.set(false);
            });

            while (active.get()) {
                Thread.sleep(50);
            }
        }
        while (this.files.size() != this.screenshots.size()) {
            Thread.sleep(50);
        }
        this.loading = false;
        refresh();
//        this.list.loadScreenshots();
    }

    /**
     * Refresh the screenshot cache.
     */
    public void refresh() {
        int selectedIndex = this.index;
        if (selectedIndex < 0) {
            this.index = selectedIndex = 0;
        }
        Screenshot selected;
        if (this.screenshots.isEmpty()) {
            selected = null;
        } else {
            if (selectedIndex >= this.screenshots.size()) {
                this.index = selectedIndex = this.screenshots.size() - 1;
            }
            selected = screenshots.get(selectedIndex);
        }
        int index;
        if (selected == null) {
            this.currentScreenshot = null;
            this.index = -1;
        } else if ((selectedIndex) != currentIndex) {
            index = selectedIndex;
            this.currentIndex = index;
            this.currentScreenshot = screenshots.get(index);
        }
    }

    @SuppressWarnings("resource")
    @Override
    public void renderBackground(@NotNull PoseStack pose) {
        fill(pose, 0, 0, width, height, 0xFF000000);
        if (currentScreenshot != null) {
            @Nullable AbstractTexture texture = currentScreenshot.texture();
            ScreenshotData data = currentScreenshot.data();
            ResourceLocation location = currentScreenshot.resourceLocation();

            if (location == null) {
                location = new ResourceLocation("");
            }

            RenderSystem.setShaderTexture(0, location);

            if (texture != null) {
                int imgWidth = data.width();
                int imgHeight = data.height();

                Resizer resizer = new Resizer(imgWidth, imgHeight);
                FloatSize size = resizer.thumbnail(this.width, this.height);

                int centerX = this.width / 2;
                int centerY = this.height / 2;
                int width = (int) size.width;
                int height = (int) size.height;

                blit(pose,  centerX - width / 2, centerY - height / 2, width, height, 0, 0, imgWidth, imgHeight, imgWidth, imgHeight);
            } else {
                blit(pose,  0, 0, width, height, 0, 0, 16, 16, 16, 16);
            }
        } else if (!this.files.isEmpty() && loading) {
            pose.pushPose();
            {
                pose.scale(2, 2, 1);
                drawCenteredString(pose, font, "Loading...", width / 4, height / 4 - 14, 0xFFFFFFFF);
            }
            pose.popPose();
            drawCenteredString(pose, font, CommonTexts.loadedScreenshots(this.loaded, this.total), width / 2, height / 2, 0xFFFFFFFF);
        } else if (this.files.isEmpty()) {
            pose.pushPose();
            {
                pose.scale(2, 2, 1);
                drawCenteredString(pose, font, CommonTexts.NO_SCREENSHOTS, width / 4, height / 4, 0xFFFFFFFF);
            }
            pose.popPose();
        } else {
            pose.pushPose();
            {
                pose.scale(2, 2, 1);
                drawCenteredString(pose, font, CommonTexts.ERROR_OCCURRED, width / 4, height / 4 - 14, 0xFFFFFFFF);
            }
            pose.popPose();
            drawCenteredString(pose, font, CommonTexts.INVALID_SCREENSHOT, width / 2, height / 2, 0xFFFFFFFF);
        }
    }

    /**
     * Load texture file into a resource location.
     *
     * @param location the resource location to read the texture into.
     * @param file     the file to read.
     * @return an instance of {@linkplain DynamicTexture} containing data of the given file.
     */
    @Nullable
    public DynamicTexture loadTexture(ResourceLocation location, File file) {
        try (InputStream input = new FileInputStream(file)) {
            NativeImage nativeImage = NativeImage.read(input);
            DynamicTexture texture = new DynamicTexture(nativeImage);

            texture.setBlurMipmap(true, false);

            Minecraft mc = Minecraft.getInstance();

            mc.getTextureManager().register(location, texture);
            return texture;
        } catch (Throwable t) {
            MainMod.LOGGER.error("Couldn't read image: {}", file.getAbsolutePath(), t);
            return null;
        }
    }

    /**
     * Go back to previous screen.
     */
    public void back() {
        // Go back to the previous screen.
        Objects.requireNonNull(this.minecraft).setScreen(this.backScreen);
    }

    @Override
    public boolean keyReleased(int keyCode, int scanCode, int modifiers) {
        if (keyCode == InputConstants.KEY_LEFT) {
            prevShot();
            return true;
        }
        if (keyCode == InputConstants.KEY_RIGHT) {
            nextShot();
            return true;
        }
        return super.keyReleased(keyCode, scanCode, modifiers);
    }

    @Override
    public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
        if (KeyboardHelper.isHoldingCtrl()) {
            if (keyCode == 61 || keyCode == 334) {
                System.out.println("Zooming in...");
                return true;
            }
            if (keyCode == 45 || keyCode == 333) {
                System.out.println("Zooming out...");
                return true;
            }
        }
        return super.keyPressed(keyCode, scanCode, modifiers);
    }

    /**
     * Go to the previous screenshot.
     */
    public void prevShot() {
        if (this.index > 0) {
            this.index--;
            this.refresh();
        }
    }

    /**
     * Go to the next screenshot.
     */
    public void nextShot() {
        if (this.index < this.files.size() - 1) {
            this.index++;
            this.refresh();
        }
    }

    @Override
    public void onClose() {
        this.back();
    }

    public List<File> getFiles() {
        return Collections.unmodifiableList(this.files);
    }

    @NotNull
    public List<Screenshot> getScreenshots() {
        return Collections.unmodifiableList(this.screenshots);
    }

    public Screen getBackScreen() {
        return backScreen;
    }

    public Screenshot getCurrentScreenshot() {
        return currentScreenshot;
    }

    public int getIndex() {
        return index;
    }

    public Thread getLoadThread() {
        return loadThread;
    }

    public int getCurrentIndex() {
        return currentIndex;
    }

    public int getTotal() {
        return total;
    }

    public int getLoaded() {
        return loaded;
    }

    public boolean isLoading() {
        return loading;
    }
}

File: src/main/java/com/ultreon/mods/screenshotmanager/client/gui/screens/ScreenshotsScreen.kt
package com.ultreon.mods.screenshotmanager.client.gui.screens

import com.badlogic.gdx.Input.Keys
import com.badlogic.gdx.graphics.Pixmap
import com.badlogic.gdx.graphics.Texture
import com.ultreon.mods.screenshotmanager.ScreenshotManagerMod
import com.ultreon.mods.screenshotmanager.client.Screenshot
import com.ultreon.mods.screenshotmanager.client.ScreenshotCache
import com.ultreon.mods.screenshotmanager.client.ScreenshotData
import com.ultreon.mods.screenshotmanager.client.gui.widgets.ToolbarButton
import com.ultreon.mods.screenshotmanager.text.CommonTexts
import com.ultreon.mods.screenshotmanager.util.KeyboardHelper
import com.ultreon.mods.screenshotmanager.util.Resizer
import dev.ultreon.quantum.client.QuantumClient
import dev.ultreon.quantum.client.gui.Renderer
import dev.ultreon.quantum.text.TextObject
import dev.ultreon.quantum.util.Identifier
import dev.ultreon.quantum.util.RgbColor
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.concurrent.thread

private val emptyId = Identifier("")

class ScreenshotsScreen(title: TextObject) : FullscreenRenderScreen(title) {
    // No getter / setter.
    private val files0: MutableList<File> = ArrayList()
    var screenshot: Screenshot? = null
        private set

    val files: List<File> get() = files0.toList()

    // Getter & setter.
    var index = 0
        private set
    var total = 0
        private set
    var loaded = false
        private set
    var isLoading = false
        private set

    init {
        reload()

        addToolbarItem(ToolbarButton(0, 0, 50, message = CommonTexts.prev) { prevShot() })
        addToolbarItem(ToolbarButton(0, 0, 50, message = CommonTexts.next) { nextShot() })
    }

    private fun reload() {
        val dir = File(QuantumClient.getGameDir().toFile(), "screenshots")

        if (dir.exists()) files0 += dir.listFiles()!!

        this.total = this.files0.size
        this.index = 0
        this.refresh()
    }

    private fun loadScreenshot() {
        val active = AtomicBoolean(true)
        if (this.files0.size == 0) return
        this.files0[this.index].let { file ->
            active.set(true)
            QuantumClient.invoke {
                val texture: Texture?
                val data: ScreenshotData
                val location: Identifier?
                val fixedFilename = file.name.lowercase().replace("[^a-z0-9/._]".toRegex(), "_")
                val id = "screenshot/$fixedFilename"
                location = ScreenshotManagerMod.id(id)
                if (!QuantumClient.get().textureManager.isTextureLoaded(location)) {
                    val dynamicTexture = loadTexture(location, file)
                    texture = dynamicTexture
                    data = ScreenshotCache.cache(file, dynamicTexture)
                } else {
                    texture = QuantumClient.get().textureManager.getTexture(location)
                    data = requireNotNull(ScreenshotCache[file]) { "Screenshot ${file.path} wasn't cached." }
                }
                val screenshot = Screenshot(file, texture, location, data)
                this.screenshot = screenshot
            }
        }
    }

    /**
     * Refresh the screenshot cache.
     */
    fun refresh() {
        thread {
            loadScreenshot()
        }
    }

    override fun renderBackground(graphics: Renderer) {
        if (client!!.world != null) {
            renderTransparentBackground(graphics)
        } else {
            renderSolidBackground(graphics)
        }


        graphics.pushMatrix()
        run {
            graphics.translate(0f, 0f, 100f)
            if (this.screenshot != null) {
                val texture = this.screenshot!!.texture
                val data = this.screenshot!!.data
                var location = this.screenshot!!.id
                if (location == null) {
                    location = emptyId
                }

                if (texture != null) {
                    val imgWidth = data.width
                    val imgHeight = data.height
                    val resizer = Resizer(imgWidth.toFloat(), imgHeight.toFloat())
                    val size = resizer.thumbnail(this.width.toFloat() - 26, this.height.toFloat() - 82)
                    val centerX = this.width / 2
                    val centerY = (this.height - 69) / 2 + 48 / 2
                    val width = size.width.toInt()
                    val height = size.height.toInt()

                    graphics.fill(centerX - width / 2 - 1, centerY - height / 2 - 1, width + 2, height + 2, RgbColor.BLACK)

                    graphics.blit(
                        location,
                        centerX - width / 2f,
                        centerY - height / 2f,
                        width.toFloat(),
                        height.toFloat(),
                        0f,
                        0f,
                        imgWidth.toFloat(),
                        imgHeight.toFloat(),
                        imgWidth,
                        imgHeight
                    )
                } else {
                    graphics.blit(location, 0f, 0f, this.width.toFloat(), this.height.toFloat(), 0f, 0f, 16f, 16f, 16, 16)
                }
            } else if (this.files0.isNotEmpty() && this.isLoading) {
                graphics.textCenter(CommonTexts.loading, 2f, this.width / 2, this.height / 2 - 14)
            } else if (this.files0.isEmpty()) {
                graphics.textCenter(
                    CommonTexts.noScreenshots,
                    2f,
                    this.width / 2,
                    this.height / 2
                )
            } else {
                graphics.textCenter(
                    CommonTexts.errorOccurred,
                    2f,
                    this.width / 2,
                    this.height / 2 - 14
                )
                graphics.textCenter(
                    CommonTexts.invalidScreenshot,
                    this.width / 2,
                    this.height / 2
                )
            }
        }
        graphics.popMatrix()
    }

    /**
     * Load texture file into a resource location.
     *
     * @param location the resource location to read the texture into.
     * @param file     the file to read.
     * @return an instance of [DynamicTexture] containing data of the given file.
     */
    fun loadTexture(location: Identifier, file: File): Texture {
        try {
            FileInputStream(file).use { input ->
                val readAllBytes = input.readAllBytes()
                val pixmap = Pixmap(readAllBytes, 0, readAllBytes.size)
                val texture = Texture(pixmap)
                texture.setFilter(Texture.TextureFilter.Linear, Texture.TextureFilter.Linear)
                val mc = QuantumClient.get()
                mc.textureManager.registerTexture(location, texture)
                return texture
            }
        } catch (exception: IOException) {
            ScreenshotManagerMod.logger.error("Couldn't read image: {}", file.absolutePath)
            throw exception
        }
    }

    override fun keyRelease(keyCode: Int): Boolean {
        if (keyCode == Keys.LEFT) {
            this.prevShot()
            return true
        }
        if (keyCode == Keys.RIGHT) {
            this.nextShot()
            return true
        }

        return super.keyRelease(keyCode)
    }

    override fun keyPress(keyCode: Int): Boolean {
        if (KeyboardHelper.isHoldingCtrl) {
            if (keyCode == 61 || keyCode == 334) {
                println("Zooming in...")
                return true
            }
            if (keyCode == 45 || keyCode == 333) {
                println("Zooming out...")
                return true
            }
        }
        return super.keyPress(keyCode)
    }

    /**
     * Go to the previous screenshot.
     */
    fun prevShot() {
        if (this.index > 0) {
            this.index--
            this.refresh()
        }
    }

    /**
     * Go to the next screenshot.
     */
    fun nextShot() {
        if (this.index < this.files0.size - 1) {
            this.index++
            this.refresh()
        }
    }

    fun open() {
        this.client.showScreen(this)
    }
}